### DISCUSSION TITLE
Object equality comparison for disjunctive patterns

### DISCUSSION THREAD

**Discussion by CypherPotato** [dotnet/csharplang#8029] [Language Ideas]
As discussed in https://github.com/dotnet/csharplang/discussions/8028#discussioncomment-8987882, we can check if a value is part of the set with the `is` operator:

```cs
int value = 20;
if (value is 20 or 30 or 40) {
    // something
}
```

but this expression does not work for non-constant values. The proposal is to check whether the value is still part of a set, through its `Equals` method or `==` operator, for each embedded value, such as:

```cs
int value = 20;
if (value == 20 or 30 or 40) {
    // something
}

// should work for any expression
if (value == other1 or other2 or other3) {
    // something
}
```

Which would be lowered to:

```cs
if (value == other1 || value == other2 || value == other3) {
    // something
}
```

or even

```cs
if (value.Equals(other1) || value.Equals(other2) || value.Equals(other3)) {
    // something
}
```

Something like this would be more syntax sugar, but it would reduce the amount of code needed to achieve the given goal.

**Comment by HaloFour** [comment]
I don't believe the `or` syntax could be used here as that is specific to disjunctive patterns, and that had to be different from `||` already to disambiguate it from Boolean expressions.  Patterns are designed to apply only to constant values and there has been no discussion to allow patterns to support variables in general.

**Reply by CypherPotato** [reply]
The mention of disjunctive patterns is to express the idea of checking whether `X == Y || Z`, without having to rewrite `X` everytime, as `X == Y || X == Z`. Also I would like you to link to the discussion that concluded that it would not be a good idea to check variables/references in this way, to help me understand it.

**Reply by HaloFour** [reply]
Here's a comment within a discussion: https://github.com/dotnet/csharplang/issues/2850#issuecomment-623226762

It doesn't rule out revisiting this in the future, just that it wasn't part of the design.

**Comment by huoyaoyuan** [comment]
It's basically distributive property on binaryOp over patternOp:

`value0 binaryOp (value1 patternOp value2 patternOp ...)`
becomes
`(value0 binaryOp value1) patternOp (value0 binaryOp value2) patternOp ...`
